File: コードレビュー/04_上級/プログラム作成問題：上級編/20190924_165353689/NameBattler/src/Player.java

// プレイヤークラス(各種ジョブの基底クラス)

import java.math.BigInteger;
import java.security.MessageDigest;
import java.util.Random;

public class Player {
	// =======================
	// フィールド変数
	// =======================
	// 名前
	protected String name;
	// HP
	protected int hp;
	//MaxHP
	protected int maxhp;
	// MP
	protected int mp;
	// 攻撃力
	protected int str;
	// 防御力
	protected int def;
	// 幸運
	protected int luck;
	//速度
	protected int agi;

	protected boolean paralyze;

	protected boolean poison;

	// =======================
	// コンストラクタ
	// =======================
	/**
	 * コンストラクタ
	 * @param name : プレイヤー名
	 */
	public Player(String name) {
		this.name = name;

		// キャラクターのパラメータ生成
		MakeCharacter();
	}

	// =======================
	// Getter / Setter
	// =======================
	/**
	 * プレイヤー名を取得する
	 * @return プレイヤー名
	 */
	public String GetName() {
		return this.name;
	}

	/**
	 * 現在HPを取得する
	 * @return 現在HP
	 */
	public int GetHP() {
		return this.hp;
	}

	/**
	 * HPの最大値を取得する
	 * @return 最大HP
	 */
	public int GetMaxHP() {
		return this.maxhp;
	}

	/**
	 * 現在MPを取得する
	 * @return 現在MP
	 */
	public int GetMP() {
		return this.mp;
	}

	/**
	 * 攻撃力を取得する
	 * @return 攻撃力
	 */
	public int GetSTR() {
		return this.str;
	}

	/**
	 * 防御力を取得する
	 * @return 防御力
	 */
	public int GetDEF() {
		return this.def;
	}

	/**
	 * 幸運値を取得する
	 * @return 幸運値
	 */
	public int GetLUCK() {
		return this.luck;
	}

	/**
	 * 速さを取得する
	 * @return 速さ
	 */
	public int GetAGI() {
		return this.agi;
	}

	/**
	 * 麻痺状態かどうかを取得する
	 * @return 麻痺状態可否
	 */
	public boolean GetParalyze() {
		return this.paralyze;
	}

	/**
	 * 毒状態かどうかを取得する
	 * @return 毒状態可否
	 */
	public boolean GetPoison() {
		return this.poison;
	}
	// =======================
	// protected メソッド
	// =======================
	/**
	 * 名前(name)からキャラクターに必要なパラメータを生成する
	 */
	protected void MakeCharacter() {
		// ジョブごとにオーバーライドして処理を記述してください
	}

	/**
	 * 名前(name)からハッシュ値を生成し、指定された位置の数値を取り出す
	 * @param index : 何番目の数値を取り出すか
	 * @param max : 最大値(内部的に0～255の値を生成するが、0～maxまでの値に補正)
	 * @return 数値(0～max) ※maxも含む
	 */
	protected int GetNumber(int index, int max) {
		try {
			// 名前からハッシュ値を生成する
			byte[] result = MessageDigest.getInstance("SHA-1").digest(this.name.getBytes());
			String digest = String.format("%040x", new BigInteger(1, result));

			// ハッシュ値から指定された位置の文字列を取り出す（２文字分）
			String hex = digest.substring(index * 2, index * 2 + 2);

			// 取り出した文字列（16進数）を数値に変換する
			int val = Integer.parseInt(hex, 16);
			return val * max / 255;
		} catch (Exception e) {
			// エラー
			e.printStackTrace();
		}
		return 0;
	}

	// =======================
	// private メソッド
	// =======================

	// =======================
	// public メソッド
	// =======================
	/**
	 * 現在のステータスを System.out で表示する
	 */
	public void PrintStatus() {
		System.out.printf("%s (HP=%3d : MP=%3d : STR=%3d : DEF=%3d LUCK=%3d : AGI=%3d)\n", this.GetName(), this.GetHP(),
				this.GetMP(), this.GetSTR(), this.GetDEF(), this.GetLUCK(), this.GetAGI());
	}

	/**
	 * 対象プレイヤー(target)に攻撃を行う
	 * @param attacker : 攻撃側プレーヤー
	 * @param defender : 防御側プレイヤー
	 */
	public void Attack(Player attacker, Player defender) {
		// ジョブごとにオーバーライドして処理を記述してください
	}

	/**
	 * 対象プレイヤー(target)に対して与えるダメージを計算する
	 * @param target : 対象プレイヤー
	 * @return ダメージ値(0～)
	 */
	protected int CalcDamage(Player target) {
		Random random = new Random();
		int critical = random.nextInt(100) + 1;

		// 会心の一撃かどうか
		if (GetLUCK() > critical) {
			System.out.println("会心の一撃！");
			return GetSTR();
		}

		int damage = GetSTR() - target.GetDEF();

		if (damage < 0) {
			damage = 0;
		}
		return damage;

	}

	/**
	 * ダメージを受ける
	 * @param damage : ダメージ値
	 */
	protected void Damage(int damage) {
		// ダメージ値分、HPを減少させる
		this.hp = Math.max(this.GetHP() - damage, 0);
	}

	/**
	* HPを回復する
	* @param damage : 回復値
	*/
	protected void RecoverHP(int recoverHP) {
		// 回復量分、HPを上昇させる
		this.hp = Math.min(this.GetHP() + recoverHP, this.maxhp);

	}

	/**
	 * MPを消費する
	 * @param usemp : 消費するMPの値
	 */
	protected void UseMP(int usemp) {

		// 消費MP分、MPを減少させる
		this.mp = Math.max(this.GetMP() - usemp, 0);
	}

	/**
	 * 麻痺状態にする
	 * @param status : 麻痺状態にする
	 */
	protected void SetParalyze(boolean status) {
		this.paralyze = status;
	}
	/**
	 * 毒状態にする
	 * @return 毒状態にする
	 */
	protected void SetPoison(boolean status) {
		this.poison = status;
	}
	/**
	 * 毒状態になっている場合の処理
	 */
	public void ProcessPoison() {
		this.Damage(20);

}




}

File: コードレビュー/07.Unityアプリ/20200701_USSBB1M39/Assets/Script/Player.cs
using System;
using System.Text;
using System.Collections;
using System.Collections.Generic;
using System.Security.Cryptography;
using UnityEngine;

public class Player
{
	// =======================
	// フィールド変数
	// =======================
	// 名前
	protected string name;
	// HP
	protected int hp;
	protected int defaultHP;
	// MP
	protected int mp;
	protected int defaultMP;
	// 攻撃力
	protected int str;

	// 防御力
	protected int def;

	// 幸運
	protected int luck;

	//速度
	protected int agi;

	// 麻痺
	protected bool paralyze;

	protected int paralyzeTurn;

	// 毒
	protected bool poison;

	//行動できるかどうか
	protected bool active;

	protected Strategy strategy;

	// 所属しているパーティー
	protected Party myParty;

	// =======================
	// コンストラクタ
	// =======================

	/// <summary>
	/// コンストラクタ
	/// </summary>
	/// <param name="name">プレイヤー名</param>
	public Player(string name)
	{
		this.name = name;
		this.strategy = new DefaultStrategy();
		// キャラクターのパラメータ生成
		MakeCharacter();
	}

	// =======================
	// Getter / Setter
	// =======================
	/// <summary>
	/// プレイヤー名を取得する
	/// </summary>
	/// <returns>プレイヤー名</returns>
	public string GetName()
	{
		return this.name;
	}

	/// <summary>
	/// 現在HPを取得する
	/// </summary>
	/// <returns> 現在HP</returns>
	public int GetHP()
	{

		return this.hp;
	}

	/// <summary>
	/// 最大HPを取得する
	/// </summary>
	/// <returns>最大HP</returns>
	public int GetDefaultHP()
	{
		return this.defaultHP;
	}

	/// <summary>
	/// 現在MPを取得する
	/// </summary>
	/// <returns>現在MP</returns>
	public int GetMP()
	{
		return this.mp;
	}

	/// <summary>
	/// 最大MPを取得する
	/// </summary>
	/// <returns>最大MP</returns>
	public int GetDefaultMP()
	{
		return this.defaultMP;
	}

	/// <summary>
	/// 攻撃力を取得する
	/// </summary>
	/// <returns>攻撃力</returns>
	public int GetSTR()
	{
		return this.str;
	}

	/// <summary>
	/// 防御力を取得する
	/// </summary>
	/// <returns></returns>
	public int GetDEF()
	{
		return this.def;
	}

	/// <summary>
	/// 幸運値を取得する
	/// </summary>
	/// <returns></returns>
	public int GetLUCK()
	{
		return this.luck;
	}

	/// <summary>
	/// 速さを取得する
	/// </summary>
	/// <returns>速さ</returns>
	public int GetAGI()
	{
		return this.agi;
	}


	/// <summary>
	/// 麻痺状態になっているかどうか
	/// </summary>
	/// <returns>true：麻痺状態</returns>
	public bool isParalyze()
	{
		return this.paralyze;
	}
	/// <summary>
	/// 行動できるかどうか
	/// </summary>
	/// <returns>true : 行動可能</returns>
	public bool isActive()
	{
		return active;
	}

	/// <summary>
	/// 毒状態になっているかどうか
	/// </summary>
	/// <returns>true : 毒状態</returns>
	public bool isPoison()
	{
		return this.poison;
	}

	/// <summary>
	/// 麻痺状態であるか設定する
	/// </summary>
	/// <param name="paralyze">true：麻痺状態<br></br>false : 麻痺状態でない</param>
	public void SetParalyze(bool paralyze)
	{
		this.paralyze = paralyze;
		this.paralyzeTurn = 5;
	}

	/// <summary>
	/// 麻痺状態が回復する
	/// </summary>
	public void RecoveryParalyze()
	{
		this.paralyze = false;
	}

	/// <summary>
	/// 麻痺状態が残り何ターン続くか
	/// </summary>
	/// <returns>残り何ターン数</returns>
	public int GetParalyzeTurn()
	{
		return this.paralyzeTurn;
	}

	/// <summary>
	/// 麻痺状態の残りターン数を減らす
	/// </summary>
	public void ChangeParalyzeTurn()
	{
		if (paralyzeTurn > 0)
		{
			this.paralyzeTurn--;
		}
	}

	/// <summary>
	/// 毒状態であるか設定する
	/// </summary>
	/// <param name="poison">true：毒状態<br></br>false : 毒状態でない</param>
	public void SetPoison(bool poison)
	{
		this.poison = poison;
	}




	/// <summary>
	/// 自分のパーティーを取得する
	/// </summary>
	/// <returns>自分のパーティー</returns>
	public Party GetMyParty()
	{
		return this.myParty;
	}


	/// <summary>
	/// 自分がどのパーティーに所属しているか設定する
	/// </summary>
	/// <param name="myParty">自身の属するパーティー</param>
	public void SetMyParty(Party myParty)
	{
		this.myParty = myParty;

	}

	/// <summary>
	/// 自分のパーティーの作戦を設定する
	/// </summary>
	/// <param name="strategy">自パーティーの作戦</param>
	public void SetStrategy(Strategy strategy)
	{
		this.strategy = strategy;
	}




	// =======================
	// protected メソッド
	// =======================
	/// <summary>
	/// 名前(name)からキャラクターに必要なパラメータを生成する<br></br>
	/// ジョブごとにオーバーライド
	/// </summary>
	protected virtual void MakeCharacter()
	{
		// ジョブごとにオーバーライドして処理を記述してください
	}

	/// <summary>
	/// 名前(name)からハッシュ値を生成し、指定された位置の数値を取り出す
	/// </summary>
	/// <param name="index">指定位置</param>
	/// <param name="max">最大値(内部的に0～255の値を生成するが、0～maxまでの値に補正)</param>
	/// <returns>数値(0～max) ※maxも含む</returns>
	protected int GetNumber(int index, int max)
	{
		Debug.Log("GetNumberが呼び出されました");
		try
		{
			Debug.Log(this.name);

			// バイト配列に
			byte[] byteValue = Encoding.UTF8.GetBytes(this.name);

			// ハッシュ値の取得
			SHA1 crypto = new SHA1CryptoServiceProvider();
			byte[] hashValue = crypto.ComputeHash(byteValue);

			StringBuilder hashedText = new StringBuilder();
			for (int i = 0; i < hashValue.Length; i++)
			{
				hashedText.AppendFormat("{0:x2}", hashValue[i]);
			}
			Debug.Log("ハッシュ値：" + hashedText.ToString());
			Debug.Log("文字数：" + hashedText.Length.ToString());

			String hex = hashedText.ToString().Substring(index * 2, 2);
			Debug.Log(hex);

			int value = Int16.Parse(hex, System.Globalization.NumberStyles.HexNumber);
			Debug.Log(value);
			return value * max / 255;
		}
		catch (Exception e)
		{
			// エラー
			Console.WriteLine("例外をキャッチしました");
			Debug.Log("例外をキャッチしました");
			Console.WriteLine(e);
		}
		return 0;
	}

	// =======================
	// private メソッド
	// =======================

	// =======================
	// public メソッド
	// =======================

	/// <summary>
	/// 現在のステータスを表示する
	/// </summary>
	/// <returns></returns>
	public string PrintStatus()
	{
		string status = "";
		status = string.Format("{0}(HP={1}/{2} : MP={3}/{4} : STR={5} : DEF={6} : LUCK={7} : AGI={8})", this.name, this.hp, this.defaultHP, this.mp, this.defaultMP, this.str, this.def, this.luck, this.agi);

		return status;
	}

	/// <summary>
	/// Playerの行動を作戦に応じて決定し実行する
	/// </summary>
	/// <param name="activePlayer"></param>
	/// <param name="enemyParty"></param>
	public void Action(Player activePlayer, Party enemyParty)
	{
		Debug.Log("Player Action ");
		this.strategy.StrategyAction(activePlayer, activePlayer.GetMyParty(), enemyParty);

	}

	/// <summary>
	/// 対象パーティー(passiveParty)に攻撃を行う
	/// </summary>
	/// <param name="activePlayer">能動側プレイヤー</param>
	/// <param name="passiveParty">受動側パーティー</param>
	public virtual void Attack(Player activePlayer, List<Player> passiveParty)
	{
		// ジョブごとにオーバーライドして処理を記述してください
	}

	/// <summary>
	/// 対象パーティー(passiveParty)に回復を行う
	/// </summary>
	/// <param name="activePlayer">能動側プレイヤー</param>
	/// <param name="passiveParty">受動側パーティー</param>
	/// <returns></returns>
	public virtual bool HealHP(Player activePlayer, List<Player> passiveParty)
	{
		// ジョブごとにオーバーライドして処理を記述してください
		return false;
	}

	public virtual bool HealDebuff(Player ActivePlayer, Player passivePlayer)
	{
		// ジョブごとにオーバーライドして処理を記述してください
		return false;
	}

	/// <summary>
	/// 対象プレイヤー(passivePlayer)に状態異常魔法を使う
	/// </summary>
	/// <param name="activePlayer">能動側プレイヤー</param>
	/// <param name="passivePlayer">受動側プレイヤー</param>
	/// <returns></returns>
	public bool Debuff(Player activePlayer, Player passivePlayer)
	{

		return false;
	}


	/// <summary>
	/// 対象プレイヤー(target)に対して与えるダメージを計算する
	/// </summary>
	/// <param name="target">対象プレイヤー</param>
	/// <returns>ダメージ値(0～)</returns>
	public int CalcDamage(Player target)
	{
		int critical = UnityEngine.Random.Range(1, 100);
		int damage;
		// 剣を持っているかどうか

		// 会心の一撃かどうか
		if (GetLUCK() > critical)
		{

			LogText.AddLog("会心の一撃！");

			return GetSTR();

		}

		damage = str - target.GetDEF();

		if (damage < 0)
		{
			return 0;
		}
		return damage;

	}

	/// <summary>
	/// ダメージを受ける
	/// </summary>
	/// <param name="damage">ダメージ値</param>
	public void Damage(int damage)
	{
		// ダメージ値分、HPを減少させる
		this.hp = Math.Max(this.hp - damage, 0);
	}


	/// <summary>
	/// HPを回復する
	/// </summary>
	/// <param name="recoverHP">回復値</param>
	public void RecoverHP(int recoverHP)
	{
		// 回復量分、HPを上昇させる
		this.hp = Math.Min(this.hp + recoverHP, this.defaultHP);

	}



	/// <summary>
	/// MPを消費する
	/// </summary>
	/// <param name="usemp"> 消費するMPの値</param>
	public void UseMP(int usemp)
	{

		// 消費MP分、MPを減少させる
		this.mp = (Math.Max(this.mp - usemp, 0));
	}


	/// <summary>
	/// 毒状態になっている場合の処理
	/// </summary>
	public void ProcessPoison()
	{
		if (this.poison)
		{
			this.Damage(20);
		}
	}


	/// <summary>
	/// PlayerのHPがゼロになったときに表示する
	/// </summary>
	public void Down()
	{
		if (this.hp == 0)
		{
			LogText.AddLog(string.Format("{0}は力尽きた...", this.name));
		}
	}

	/// <summary>
	/// 行動できる状態にあるか確認する
	/// </summary>
	/// <returns>true：行動可能<br></br>false : 行動不能</returns>
	public bool GetActive()
	{
		return this.active;
	}


	public void ChangeActive(bool active)
	{
		this.active = active;
	}

	/// <summary>
	/// ステータスを初期状態に戻す
	/// </summary>
	public void ResetStatus()
	{
		this.hp = this.defaultHP;
		this.mp = this.defaultMP;
		this.paralyze = false;
		this.paralyzeTurn = 0;
		this.poison = false;

		this.active = true;
	}
}